client-misk-spanner/src/main/kotlin/app/cash/backfila/client/misk/spanner/internal/SpannerBackfillOperator.kt

```diff
@@ -0,0 +1,108 @@
+package app.cash.backfila.client.misk.spanner.internal
+
+import app.cash.backfila.client.misk.spanner.SpannerBackfill
+import app.cash.backfila.client.spi.BackfillOperator
+import app.cash.backfila.client.spi.BackfilaParametersOperator
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeRequest
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse
+import app.cash.backfila.protos.clientservice.PrepareBackfillRequest
+import app.cash.backfila.protos.clientservice.PrepareBackfillResponse
+import app.cash.backfila.protos.clientservice.RunBatchRequest
+import app.cash.backfila.protos.clientservice.RunBatchResponse
+import com.google.cloud.spanner.Key
+import com.google.cloud.spanner.KeyRange
+import com.google.cloud.spanner.KeySet
+import com.squareup.moshi.Moshi
+import misk.moshi.adapter
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+
+class SpannerBackfillOperator<Param : Any> internal constructor(
+  override val backfill: SpannerBackfill<Param>,
+  private val parametersOperator: BackfilaParametersOperator<Param>,
+  backend: SpannerBackend,
+) : BackfillOperator {
+  private var moshi: Moshi = backend.moshi
+  private val adapter = moshi.adapter<List<String>>()
+
+  override fun name() = backfill.javaClass.toString()
+
+  override fun prepareBackfill(request: PrepareBackfillRequest): PrepareBackfillResponse {
+    val config =
+      parametersOperator.constructBackfillConfig(request.parameters, request.dry_run)
+    backfill.validate(config)
+
+    val partitions = listOf(
+      PrepareBackfillResponse.Partition.Builder()
+        .backfill_range(request.range)
+        .partition_name("partition")
+        .build()
+    )
+
+    return PrepareBackfillResponse.Builder()
+      .partitions(partitions)
+      .build()
+  }
+
+  override fun getNextBatchRange(request: GetNextBatchRangeRequest): GetNextBatchRangeResponse {
+    val range = if (request.previous_end_key == null) {
+      KeySet.all()
+    } else {
+      val previousEndKey = adapter.fromJson(request.previous_end_key.utf8())!!
+      KeySet.range(
+        KeyRange.openClosed(
+          Key.of(*previousEndKey.toTypedArray()),
+          Key.of(),
+        )
+      )
+    }
+
+    val query = backfill.dbClient.singleUseReadOnlyTransaction()
+      .read(backfill.tableName, range, backfill.primaryKeyColumns)
```

Comment:
I believe you can pass options to this query. You should make it a "low priority" query - otherwise a backfill which runs at "max speed" could overwhelm the DB. Plus we want to tell Spanner that backfill traffic isn't as important.

Reply:
As far as I can tell, options can only be passed for transactions that can perform writes.